dialect: (irdl) Add region and attributes operations

alexarice opened the pull request:
Adds `RegionsOp`, `RegionOp`, `RegionType`, and `AttributesOp` to irdl

superlopuh commented on tests/filecheck/dialects/irdl/testd.irdl.mlir:
```diff
@@ -144,5 +144,39 @@ builtin.module {
     }
+
+    // CHECK:      irdl.operation @op_with_regions {
+    // CHECK-NEXT:   %{{.*}} = irdl.region
+    // CHECK-NEXT:   %{{.*}} = irdl.region()
+    // CHECK-NEXT:   %{{.*}} = irdl.is i32
+    // CHECK-NEXT:   %{{.*}} = irdl.is i64
+    // CHECK-NEXT:   %{{.*}} = irdl.region(%{{.*}}, %{{.*}})
+    // CHECK-NEXT:   %{{.*}} = irdl.region with size 3
+    // CHECK-NEXT:   irdl.regions(%{{.*}}, %{{.*}}, %{{.*}}, %{{.*}})
```
why anonymize?

alexarice replied:
thought it was the done thing

superlopuh replied:
Yes and no, we don't want spurious changes in the filecheck tests if we add something in the middle, but if you print the IR you'll see that the names of the values are preserved, so it will be stable, and they're actually really useful in this case to track how the constraints are propagated, I'd recommend deanonymizing. I added an option to `filecheckize` to do this: `--xdsl-anonymize`, this will only anonymize `%[\d]+`, not ssa values with name hints.

superlopuh replied:
`cat my_ir.mlir | filecheckize`

alexarice replied:
and where would one get filecheckize from?

superlopuh replied:
`pip`, as usual :)

alexarice replied:
Ah I was trying to find it as a subproject of xdsl

superlopuh commented on tests/filecheck/dialects/irdl/testd.irdl.mlir:
```diff
@@ -144,5 +144,39 @@ builtin.module {
+      %0 = irdl.any
+      %1 = irdl.is i64
+      irdl.attributes {
+        "attr1" = %0,
+        "attr2" = %1
```
how does this work?

superlopuh replied:
Oh I see it's fully custom syntax

alexarice replied:
yes, I also wasn't sure if it should be printed in this way or if the single line print is acceptable?

superlopuh replied:
I think this looks nicer

alexarice replied:
Ok I'll try to figure out how new lines/indentation work

superlopuh commented on xdsl/dialects/irdl/irdl.py:
```diff
@@ -379,6 +394,64 @@ def print(self, printer: Printer) -> None:
+        printer.print(" ")
+        printer.print_dictionary(
+            dictionary,
+            lambda x: printer.print_attribute(x),
+            lambda x: printer.print_operand(x),
```
does this work?
```suggestion
            printer.print_attribute,
            printer.print_operand,
```

alexarice replied:
Just checked, seems to work

superlopuh commented on xdsl/dialects/irdl/irdl.py:
```diff
@@ -379,6 +394,64 @@ def print(self, printer: Printer) -> None:
+
+    attribute_value_names = attr_def(ArrayAttr[StringAttr])
+
+    def __init__(self, attributes: dict[str, SSAValue]):
+        operands = tuple(attributes.values())
+        names = ArrayAttr(StringAttr(x) for x in attributes.keys())
+        super().__init__(
+            operands=(operands,), attributes={"attribute_value_names": names}
+        )
```
I would recommend adding a dumb init with the fields of the op, and a `.get` helper with the smarts, similarly below, there's currently no way to create these ops with exactly the things you want

alexarice replied:
Is there example of such a `get` helper somewhere?

superlopuh replied:
Not sure, I just mean a class/static method that returns the Operation

superlopuh commented on xdsl/printer.py:
```diff
@@ -259,12 +259,17 @@ def print_dictionary(
-            print_key(key)
-            self.print_string("=")
-            print_value(value)
+        if not elems:
+            self.print("{}")
+        else:
+            self.print("{ ")
+            for i, (key, value) in enumerate(elems.items()):
+                if i:
+                    self.print_string(delimiter)
+                print_key(key)
+                self.print_string(" = ")
+                print_value(value)
+            self.print(" }")
```
Just noticed this, I don't think it belongs in a PR tagged as `dialects`, since it's actually a core change. Why is it necessary?

alexarice replied:
It's not "necessary", I could just put this logic in the printer for AttributesOp, but I thought it would make sense to use the print_dictionary function and I (maybe incorrectly) thought that since it was used nowhere else that it would be safe to modify/improve

alexarice replied:
Would be happy to split this change off into a separate pr if wanted

superlopuh replied:
Hmm, that's weird that it's not used... I'm also surprised none of the tests fail... My first instinct is to remove the helper if it's not used. In either case, I would probably recommend do this in a custom way in your custom syntax.

alexarice replied:
I have reverted the core change and put improved printing in `AttributesOp`

alexarice commented:
I simplified the printing a bit, removing the empty case as this should never occur in practice (in fact mlir does not even parse it). I also added the verifier that mlir has to check that there are the same number of names and values, is it worth adding a test for this?

alexarice commented:
Just realised that mlir does parse the empty case, but does it by omitting the curly braces entirely

superlopuh commented on tests/filecheck/dialects/irdl/testd.irdl.mlir:
```diff
@@ -144,5 +144,49 @@ builtin.module {
+    }
+
+    // CHECK:      irdl.operation @attr_op {
+    // CHECK-NEXT:   %{{.*}} = irdl.any
+    // CHECK-NEXT:   %{{.*}} = irdl.is i64
+    // CHECK-NEXT:   irdl.attributes {
+    // CHECK-NEXT:     "attr1" = %{{.*}},
+    // CHECK-NEXT:     "attr2" = %{{.*}}
```
can you deanon this, please?

alexarice replied:
Is this deanonification sufficient?

superlopuh replied:
it's perfect, thank you!
